parsers/dortmund.py

```diff
@@ -0,0 +1,138 @@
+import re
+import datetime
+import urllib.request as rq
+
+from urllib.parse import urlencode as urlencode
+from bs4 import BeautifulSoup
+from utils import Parser
+
+from pyopenmensa.feed import LazyBuilder
+
+# dictionary for categories
+categories = {
+    103: 'Tagesgericht',
+    101: 'Menü 1',
+    102: 'Menü 2',
+    104: 'Vegetarisches Menü',
+    105: 'Aktionsteller',
+    106: 'Aktionsteller Fisch',
+    107: 'Aktionsteller Vegan',
+    108: 'Grillstation',
+    109: 'Selfservice Station',
+    115: 'Restaurant',
+    116: 'Beilage',
+    120: 'Menü 1',
+    121: 'Menü 2',
+    122: 'Menü 3',
+    123: 'Menü 3',
+    124: 'Vegetarisches Menü',
+    125: 'Vegan',
+    127: 'Aktionsteller',
+    128: 'Schnelle Theke',
+    130: 'Sonstige',
+    131: 'Low Carb Teller',
+    135: 'Pastatheke Pastaauswahl',
+    136: 'Pastatheke Toppings',
+    137: 'Pastatheke Saucen',
+}
+
+def getAndFormatPrice(price):
+    price = re.search('(\d+,\d{1,2})', price)
+    if price != None:
+        formatted = re.sub('(\d+),(\d+)', r'\1.\2', price.group(0))
+        return float(formatted)
+    else:
+        return '-' 
+
+def getWeekdays(day):
+    currentWeekday = day.weekday()
+    week = []
+    wstart = -currentWeekday
+    wend = 7 - currentWeekday
+    for d in range(wstart, wend):
+        nday = day + datetime.timedelta(days=d)
+        week.append(nday.isoformat())
+    return week
+
+def parse_url(url, today=False):
+    canteen = LazyBuilder()
+    day = datetime.date.today()
+
+    week = getWeekdays(day)
+
+    for d in range(7):
+        py = {'tx_pamensa_mensa[date]' : week[d]}
+        payload = urlencode(py).encode('ascii')
+        data = rq.urlopen(url, payload).read().decode('utf-8')
+        soup = BeautifulSoup(data, 'html.parser')
+
+        parse_day(canteen, soup, week[d])
+
+    legend_url = 'https://www.stwdo.de/mensa-co/allgemein/zusatzstoffe/'
+    canteen.legend = parse_legend(legend_url)
+
+    return canteen.toXMLFeed()
+
+def parse_legend(url):
+    data = rq.urlopen(url).read().decode('utf-8')
+    soup = BeautifulSoup(data, 'html.parser')
+
+    table = soup.find('table', { 'class' : 'ce-table' })
+    tbody = soup.find('tbody')
+
+    legend = {}
+
+    for tr in tbody.find_all('tr'):
+        td = tr.find_all('td')
+        legend[td[0].text] = td[1].text
+
+    return legend
+
+def parse_day(canteen, soup, wdate):
+    global categories
+    mealsBody = soup.find('div', { 'class' : 'meals-body' })
+
+    for meal in mealsBody.find_all('div', { 'class' : 'meal-item' }):
+        for item in meal.find_all('div', { 'class' : 'item' }):
+            cl1 = item['class'][1]
+            if cl1 == 'category':
```

Comment:
This is rather brittle and will break if the order of the classes changes, which has no semantic meaning. You could replace this with `if 'category' in item['class']:`.

Reply:
Ok I replaced it in the whole function:

```
for meal in mealsBody.find_all('div', { 'class' : 'meal-item' }):
         for item in meal.find_all('div', { 'class' : 'item' }):
             if 'category' in item['class']:
                 img = item.find('img')
                 category = define_category(item, img)
             elif 'description' in item['class']:
                 description = item.text.strip()
             elif 'supplies'in item['class']:
                 supplies = []
                 for supply in item.find_all('img'):
                     if supply['title']:
                         supplies.append(supply['title'])
             elif 'price'in item['class']:
                 price = item.text
                 if 'student' in item['class']:
                     student_price = getAndFormatPrice(price)
                 elif 'staff' in item['class']:
                     staff_price = getAndFormatPrice(price)
                 elif 'guest' in item['class']:
                     guest_price = getAndFormatPrice(price)
```